### ISSUE TITLE
Enhancement: Add a way to hook into `onCancel` and `onConfirm`

### ISSUE THREAD

**Issue by MilesBHuff** [sshyam-gupta/react-router-prompt#95]
I would like to be able to either hook into `onCancel`/`onConfirm` _or_ change what they do.
Is this possible in the current version of the library?  If not, is this something you would be interested in adding?
Thanks.

**Comment by sshyam-gupta** [comment]
@MilesBHuff can you explain what you are trying to achieve here?

One way to do above is in your click event handler you write your logic and then call these methods, are you suggesting library should accept those functions, execute them (before) and then call core methods (onCancel/onConfirm)?

basically props as `beforeCancel`, `beforeConfirm`?

**Comment by MilesBHuff** [comment]
Yeah, that's my suggestion!

The business wanted a modal where they could choose to save/discard/cancel.  Calling "onConfirm" from inside a function, as you suggest, results in errors (I'm not at my work computer, presently, but I can provide you with the exact error next week).

EDIT:  My issue is literally identical to @jornncc's below.

**Comment by jornncc** [comment]
I have pretty much the same problem/situation - our current ui shows a 'you have unsaved changes' - with the following modal options:

1. Save (and navigate where you intended)
2. Discard Changes (and navigate where you intended)
3. Cancel (and stay where you are)

Options 2 and 3 work fine, its the 1st 'save and then call onConfirm' where a problem occurs:

```
<ReactRouterPrompt when={_isDirty}>
  {({ isActive, onConfirm, onCancel }) => (
    <SaveChangesDialog
      onSave={() => {
        _onSubmitCustomer().then(() => { onConfirm(); });
      }}
      onNo={() => onConfirm()}
      onClose={() => onCancel()}
    />
  )}
</ReactRouterPrompt>
```

_onSubmitCustomer() is the function I call to save the customer object, and it returns a promise which resolves when saving is done - then I attempt to call onConfirm() to tell the prompt to continue navigating where the end user intended... 

This ends up throwing the following error and the app fails:

```
caught Error: Invalid blocker state transition: unblocked -> proceeding
    at invariant (router.js:270:1)
    at updateBlocker (router.js:2739:1)
    at proceed (router.js:1769:1)
    at sentryWrapped (helpers.js:90:1)
i

```

**Comment by satishdeshbhratar07** [comment]
Any updates on the mentioned issue?? I am facing the same problem.

**Comment by sshyam-gupta** [comment]
I will check and resolve this issue in the weekend.
Sorry for keeping you all waiting.

**Comment by sshyam-gupta** [comment]
@satishdeshbhratar07 / @jornncc / @MilesBHuff 
I tried it here with latest version https://codesandbox.io/s/react-router-prompt-example-react-router-6-7-with-delay-mg097k?file=/src/Form.js it works fine. 
Used mock delay, can you please confirm the package version you are using? Or better if you can create sandbox for me that would help me more in investigating?

**Comment by satishdeshbhratar07** [comment]
Hi @sshyam-gupta, I am using react-router-dom v6.8 and react-router-prompt v0.5.2. I have tried your sandbox code, I am calling my own function to save the data before navigating to another page but the issue still persists.

**Comment by jornncc** [comment]
> @satishdeshbhratar07 / @jornncc / @MilesBHuff I tried it here with latest version https://codesandbox.io/s/react-router-prompt-example-react-router-6-7-with-delay-mg097k?file=/src/Form.js it works fine. Used mock delay, can you please confirm the package version you are using? Or better if you can create sandbox for me that would help me more in investigating?

The version we are using of react-router-prompt is 0.5.2

**Comment by sshyam-gupta** [comment]
I will add prop to handle `beforeCancel` and `beforeConfirm`.

**Comment by sshyam-gupta** [comment]
Released in v0.5.3.
Added support for `beforeCancel` and `beforeConfirm`.
Updated example [here](https://codesandbox.io/s/react-router-prompt-example-react-router-6-7-y9ug7z?file=/src/Form.js) 
Let me know if its good to close the issue.
